endpoints/data_objects/src/main.cpp

```diff
@@ -353,84 +463,210 @@ namespace
-					res.result(http::status::bad_request);
+				if (!fs::client::is_data_object(status)) {
+					log::error(
+						"{}: Logical path [{}] does not point to a data object or does not exist.",
+						fn,
+						lpath_iter->second);
+					res.result(http::status::internal_server_error);
 					res.prepare_payload();
 					return _sess_ptr->send(std::move(res));
 				}
 
-				auto iter = _args.find("offset");
-				if (iter != std::end(_args)) {
+				std::int64_t offset = 0;
+				if (const auto iter = _args.find("offset"); iter != std::end(_args)) {
 					try {
-						in.seekg(std::stoll(iter->second));
+						offset = std::stoll(iter->second);
 					}
 					catch (const std::exception& e) {
+						log::error("{}: Invalid value for [offset] parameter. Received [{}].", fn, iter->second);
+						res.result(http::status::internal_server_error);
+						res.prepare_payload();
+						return _sess_ptr->send(std::move(res));
+					}
+
+					if (std::cmp_less(offset, 0)) {
 						log::error(
-							"{}: Could not seek to position [{}] in data object [{}].",
+							"{}: Invalid value for [offset] parameter. Must be greater than or equal to 0. Received "
+							"[{}].",
 							fn,
-							iter->second,
-							lpath_iter->second);
+							iter->second);
 						res.result(http::status::bad_request);
 						res.prepare_payload();
 						return _sess_ptr->send(std::move(res));
 					}
 				}
 
-				std::vector<char> buffer;
+				const auto data_object_size = fs::client::data_object_size(conn, lpath_iter->second);
+				std::int64_t count = data_object_size - offset;
 
-				iter = _args.find("count");
-				if (iter == std::end(_args)) {
-					res.result(http::status::bad_request);
-					res.prepare_payload();
-					return _sess_ptr->send(std::move(res));
+				if (const auto iter = _args.find("count"); iter != std::end(_args)) {
+					try {
+						count = std::stoll(iter->second);
+					}
+					catch (const std::exception& e) {
+						log::error("{}: Invalid value for [count] parameter. Received [{}].", fn, iter->second);
+						res.result(http::status::internal_server_error);
+						res.prepare_payload();
+						return _sess_ptr->send(std::move(res));
+					}
+
+					if (std::cmp_less(count, 0)) {
+						log::error(
+							"{}: Invalid value for [count] parameter. Must be greater than or equal to 0. Received "
+							"[{}].",
+							fn,
+							iter->second);
+						res.result(http::status::bad_request);
+						res.prepare_payload();
+						return _sess_ptr->send(std::move(res));
+					}
+
+					if (const auto real_count = data_object_size - offset; std::cmp_greater(count, real_count)) {
+						count = real_count;
+					}
 				}
 
-				try {
-					const auto count = std::stoi(iter->second);
+				// Determine whether the data object exceeds the size of the internal buffer
+				// used for reading. If it does, create a dedicated iRODS connection and use that
+				// to stream the contents. Otherwise, we can use the pooled connection we have.
```

Comment:
Is this saying that the dedicated connection is required because it must be done in parallel? Just trying to clarify for my own understanding.

Reply:
I was trying to describe this situation ...

When the internal buffer size is exceeded, we have to perform the reads using the background thread pool. Each read would be posted to the thread pool individually and sequentially. For that reason, the reads must have a dedicated connection. We can't use the connections from the connection pool because doing that can lead to an unresponsive server.

Does that make sense?

Reply:
Will update the comment.

Reply:
Oh, sorry, missed this. I think what you wrote in the GitHub comment above would be good to include in the code comment, yes. Thanks

Reply:
Rubber ducking really does work lol.

Will update.

Reply:
Done.